src/telemetry_registry.erl

```diff
@@ -89,51 +45,35 @@ event_spans(Events) ->
                         Acc
     end, {Events, #{}}, Events).
 
-handle_call(_Request, _From, State) ->
-    {reply, ok, State}.
+-spec discover_all() -> ok.
+discover_all() ->
+    discover_all([App || {App, _, _} <- application:loaded_applications()]).
 
-handle_cast(_Msg, State) ->
-    {noreply, State}.
+-spec discover_all(application() | [application()]) -> ok.
+discover_all(Applications) ->
+    Events = maps:keys(maps:from_list([{Event, []} || Event <- gather_all_events(Applications)])),
+    persistent_term:put(?KEY, Events).
 
-handle_info(_Info, State) ->
-    {noreply, State}.
-
-terminate(_Reason, _State) ->
-    ok.
-
-code_change(_OldVsn, State, _Extra) ->
-    {ok, State}.
-
-discover_all(Application) ->
-    ok = register_application_events(Application),
+gather_all_events(Applications) when is_list(Applications) ->
+    lists:flatmap(fun gather_all_events/1, Applications);
+gather_all_events(Application) ->
     Applications = child_applications(Application),
-    ok = lists:foreach(fun discover_all/1, Applications).
+    lists:flatmap(fun gather_application_events/1, [Application | Applications]).
 
 child_applications(Application) ->
     {ok, ChildApplications} = application:get_key(Application, applications),
     {ok, ChildIncludedApplications} = application:get_key(Application, included_applications),
-    lists:flatten([ChildApplications, ChildIncludedApplications]).
-
-register_application_events(Application) ->
-    {ok, Modules} = application:get_key(Application, modules),
-    lists:foreach(fun (Module) ->
-                          register_module_events(Module, Application)
-                  end,
-                  Modules).
+    ChildApplications ++ ChildIncludedApplications.
 
-register_module_events(Module, _Application) ->
-    Events = events_for_module(Module),
-    lists:foreach(fun (Event) ->
-                          telemetry_registry_table:insert(Event, Module)
-                  end,
-                  Events).
-
-%% {event, application, module, meta}
+gather_application_events(Application) ->
+    case application:get_key(Application, modules) of
+        {ok, Modules} -> lists:flatmap(fun events_for_module/1, Modules);
+        _ -> []
+    end.
 
 -spec events_for_module(atom()) -> [telemetry:event_name()].
 events_for_module(Module) ->
     Info = Module:module_info(),
```

Comment:
This is not part of this PR but it is important part of the discussion: this won't work for Nerves . In Nerves, there are some modules you don't want to load on the host machine, because they contain NIF code that won't work locally. So preloading all modules is usually a no-no.

Maybe a better option that would still be application centric is to ask applications to store in their app environment under a key, such as `telemetry_events` all of the events, and we can traverse that simply by loading applications.

Reply:
In theory we could use `beam_lib` for not loaded modules (for loaded one it is better to use `module_info/0` as runtime-generated modules will not have BEAM files).

Reply:
Alternatively just use `code:all_loaded/0` and ignore applications at all.

Reply:
`beam_lib` is likely too expensive. It depends on the frequency this will run - so it will certainly require disclaimers. `code:all_loaded` will not traverse all modules.

Reply:
I assume that this is meant to be called once per release, so I do not think that “heaviness” of `beam_lib` will be much of the problem there.

Reply:
@hauleth this would likely just be called one time at runtime, so I believe that should also address @josevalim's concerns for Nerves.